tests/references/generate_references.py

```diff
@@ -0,0 +1,175 @@
+        gen_function = globals()[gen_function_name]
+
+    output = gen_function(inputs)
+    return gen_path, output
+
+
+def get_generator_info(dictionary: dict, gen_path: tuple) -> dict:
+    """
+    Retrieves the generator information from the dictionary using the given path.
+
+    Args:
+        dictionary: The dictionary containing generator information.
+        gen_path: A tuple representing the path to the generator in the dictionary.
+
+    Returns:
+        A dictionary containing the generator information.
+    """
+    return reduce(operator.getitem, gen_path, dictionary)
+
+
+def flatten_nested_dict(nested_dict: dict) -> dict:
+    """
+    Flattens a nested dictionary by removing the first level of nesting.
+
+    Args:
+        nested_dict: A dictionary where the values are also dictionaries.
+
+    Returns:
+        A flattened dictionary containing all key-value pairs from the input dictionary.
+    """
+    return dict(
+        chain.from_iterable(outer_val.items() for outer_val in nested_dict.values())
+    )
+
+
+def rename_gen_to_test(data: dict) -> dict:
+    """
+    Recursively traverses a dictionary and replaces "gen_" with "test_" in string keys and values.
+
+    Args:
+        data: The dictionary to be traversed and modified.
+
+    Returns:
+        The modified dictionary with "gen_" replaced by "test_".
+    """
+
+    def replace_gen_with_test(element: str) -> str:
+        return "test_" + element[4:] if element.startswith("gen_") else element
+
+    if isinstance(data, dict):
+        return {
+            replace_gen_with_test(key): (
+                rename_gen_to_test(value)
+                if isinstance(value, dict)
+                else replace_gen_with_test(value) if isinstance(value, str) else value
+            )
+            for key, value in data.items()
+        }
+
+
+def main():
+    """
+    Main function to process reference inputs, execute generators, and save results.
+
+    This function performs the following steps:
+    1. Extracts generator paths and inputs from REFERENCES_INPUTS.
+    2. Executes generators with their respective inputs.
+    3. Updates the reference data with the generator outputs.
+    4. Processes the reference data (flattening and renaming).
+    5. Serializes and validates the processed data.
+    6. Saves the processed data to a YAML file.
+    """
+    reference_data = REFERENCES_INPUTS.copy()
+
+    gen_paths_and_inputs = extract_generator_paths_and_inputs(REFERENCES_INPUTS)
+    gen_paths_and_inputs_with_outputs = starmap(
+        import_and_execute_generator, gen_paths_and_inputs
+    )
+
+    for gen_path, outputs in gen_paths_and_inputs_with_outputs:
+        gen_info = cast(dict, get_generator_info(reference_data, gen_path))
+        del gen_info["inputs"]
+        gen_info["inputs_with_expected_outputs"] = outputs
+
+    reference_data = flatten_nested_dict(reference_data)
+    reference_data = rename_gen_to_test(reference_data)
```

Comment:
For what do we need the "gen_[...]" and test_[...]" distinction in the dictionary? Why not just remove that layer, the geneneration then searches for the "inputs" and generates the "outputs", the test step later searches for both and compares against our estimator.

Reply:
There are a couple of reasons to that:

1. What we are describing in the dictionary is the full path, from the `cryptographic_estimators/tests/` directory, to the file containing the generator. We need that path to check if the file has a `.sage` or a `.py` extension, and then define which import method we need to use.

2. Generators have a completely different logic than tests. So we need to be able to describe, without ambiguity, if we will use the `sdfq_lee_brickell` **generator** function, or the `sdfq_lee_brickell` **test** function. That also applies to the files containing the functions, otherwise we will end having two files with the same name, one in validations and one in references. Easiest way I found to create that distinction was with the `gen` and `test` prefix.

3.  With that said, it's perfectly possible to remove the `gen` prefix on the dictionary and just append the `gen_` and `test_` prefix internally in the `generate_references.py` logic if you think is clearer.

About why not remove the layer, I'm not sure about what layer are you referring to, as the process you have described is very similar to the current one

1. The `generate_references.py` search, not for the inputs, but for the file and generator function in that file.
2.  We then generate the outputs after having imported our generator function.
3. Outputs are serialized, with inputs, in the references `.yaml` file.
4. The test_data (inputs + expected_outputs) are then passed to each test to compare against our estimators.

Maybe I'm not understanding what do you mean?

Reply:
I am just talking about the structure of the dictionary that we are generating. I don't think the dictionary should represent the code logic. The dictionary should just hold the test case data in the most intuitive form, which for me would be something like dict["Estimator"]["Algorithm"]["data"]=[(input1,output1),...] (or something similar).

Reply:
I, however, think that the dictionary structure should signalize as much as it can about where the information is being used. If we eliminate that prefix from the `REFERENCES_INPUT` dictionary, what prevents the test developer (at the code level, because of course the documentation is clear on that) to think that those inputs will be passed directly to the `test` writen in `validations`? The prefixes are helping to understand, on each step of the data flowing, what is the purpose of that data and who will consume it.

Reply:
Maybe I am misunderstanding the whole framework.
From my understanding there should be two things:
1. Functions to generate the dictionary for each estimator
2. Test functions that run KAT tests against the generated dictionary

But from what you write there is more than that? I suggest we clarify this on a conceptional level first, then talk about the technicalities.

Reply:
@Memphisd Sorry for the delay.

  Thinking about your last response, I realize you enlightened a critical point: my previous documentation was not good. We need to ensure that any contributor understands the whys and hows of the new framework by just reading the documentation, and that's clearly not happening.

  I have made a complete revamp of the documentation to fix this, so please review the latest commit at your convenience, and to ensure clarity, let's continue this discussion in a new thread.

  Lastly, I would appreciate more comprehensive descriptions of your concerns/suggestions in the future. I had spent a considerable amount of time writing detailed responses to your comments (and probably you do it too), just to find two weeks later that we, sadly, were not even in the same page of the discussion. I really appreciate your feedback, and it would be great if we could understand each other faster next time!